tpmstream.io: add support for swtpm log file format

berrange opened the pull request:
The swtpm program that is used to provide QEMU/KVM virtual machines with a virtual TPM can save a log of all TPM command packets.

The file format looks like this:

   Ctrl Cmd: length 4
   00 00 00 10
   Ctrl Rsp: length 4
   00 00 00 00
   SWTPM_IO_Read: length 10
   80 01 00 00 00 0A 00 00 01 81
   SWTPM_IO_Write: length 10
   80 01 00 00 00 0A 00 00 01 01
   Ctrl Cmd: length 4
   00 00 00 01
   Ctrl Rsp: length 8
   00 00 00 00 00 01 FF FF
   SWTPM_IO_Read: length 12
   80 01 00 00 00 0C 00 00 01 44 00 00
   SWTPM_IO_Write: length 10
   80 01 00 00 00 0A 00 00 00 00
   SWTPM_IO_Read: length 22
   80 01 00 00 00 16 00 00 01 7A 00 00 00 05 00 00
   00 00 00 00 00 01
   SWTPM_IO_Write: length 43
   80 01 00 00 00 2B 00 00 00 00 00 00 00 00 05 00
   ....

"Ctrl Cmd" and "Ctrl Rsp" are markers for messages on SWTPM's control channel, followed by data, which should be ignored.

"SWTPM_IO_Read" and "SWTPM_IO_WRITE" are markers for TPM commands and responses respectively, where we can capture the following data and convert to binary.

This allows viewing the swtpm command stream using

 # tpmstream convert swtpm.log

the auto format detection looks for " Ctrl Cmd:" in the leading bytes. The format can also be requested explicitly

 # tpmstream convert --in swtpm-log swtpm.log

Fixes: #7

HarryR commented:
Awesome, qemu's TPM support is a common testbed, and being able to easily debug newly developing firmware will hopefully make peoples lives easier. I second this PR.

Also, tpmstream is really useful.

joholl commented:
First up - sorry for the late reply and thank you for contributing!

Ah, parsing logs. This will surely bite me in the ass, but I'm willing to consider it.


### FYI
tpmstream can parse pcapng files. That means, if you use `swtpm socket`, you could do something like this:

```sh
swtpm socket --tpm2 -p 2321 --ctrl type=tcp,port=2322 --log fd=1,level=5 --flags not-need-init --tpmstate dir=/tmp
```

```sh
$ tshark -i lo -f "port 2321"
Capturing on 'Loopback: lo'
 ** (tshark:2165370) 08:47:55.364018 [Main MESSAGE] -- Capture started.
 ** (tshark:2165370) 08:47:55.364260 [Main MESSAGE] -- File: "/tmp/wireshark_loLSXOS2.pcapng"
 ...
```

```sh
tpmstream co --in=pcapng /tmp/wireshark_loLSXOS2.pcapng
````

### Review

Two issues: please remove the auto-detect. Not many people will expect tpmstream to auto-detect the swtpm-log format and this breaks parsing small blobs like `80 01`. Please also add at least a smoke test.

joholl commented on src/tpmstream/io/auto/marshal.py:
```diff
@@ -12,7 +13,13 @@ def detect_format_and_yield_buffer(buffer, strict=True):
     look_ahead = b""
     try:
-        look_ahead += bytes((next(buffer_iter), next(buffer_iter)))
+        look_ahead += bytes((
```
This breaks parsing small blobs like `80 01`.

I'd say let's remove the autodetection for now.

berrange replied:
Ok, fine with me.

berrange replied:
I've removed this in the new version posted

joholl commented on src/tpmstream/io/swtpm_log/marshal.py:
```diff
@@ -0,0 +1,122 @@
+                i = int(value, 16)
+                value = bytes()
+                state = STATE_WANT_HIGH_NIBBLE
+                yield i
```
This is not trivial. Please add at least one smoke test :)

berrange replied:
I've added a unit test using two example log files I've got. The full log was several MB which I figure you probably don't want in git history, so I trimmed down to a few 10's of KB, so we have something representative but not too large.

joholl commented:
[swtpm.log](https://github.com/user-attachments/files/16671110/swtpm.log)

For your convenience, this is running the `tpm2-tss` test suite against swtpm. Parsing still breaks at some point, but that might very well be another tpmstream bug. Thought it might be helpful.

berrange commented:
> First up - sorry for the late reply and thank you for contributing!
> 
> Ah, parsing logs. This will surely bite me in the ass, but I'm willing to consider it.
> ### FYI
> 
> tpmstream can parse pcapng files. That means, if you use `swtpm socket`, you could do something like this:
> 
> ```shell
> swtpm socket --tpm2 -p 2321 --ctrl type=tcp,port=2322 --log fd=1,level=5 --flags not-need-init --tpmstate dir=/tmp
> ```

My use case for this is that we're using swtpm to provide a virtual TPM for KVM virtual machines, managed by libvirt. The user does not have any mechanism for injecting arbitrary command line arguments to swtpm, and even if that were possible, in a production deployment many would not be allowed to make such a change. It is feasible, however, to tell users to turn on logging in the swtpm process, which is just a boolean flag they have to toggle and thus low risk of breaking a production deployment.

So basically when I'm dealing with support tickets from people using a TPM in their VM, I'd like to be able to tell them to toggle debugging and send me the swtpm logfile which I can then analyse with tpmstream.

> ### Review
> 
> Two issues: please remove the auto-detect. Not many people will expect tpmstream to auto-detect the swtpm-log format and this breaks parsing small blobs like `80 01`. Please also add at least a smoke test.

Sure will make both those changes.

berrange commented:
> [swtpm.log](https://github.com/user-attachments/files/16671110/swtpm.log)
>
> For your convenience, this is running the tpm2-tss test suite against swtpm. Parsing still breaks at some point, but that might very well be another tpmstream bug. Thought it might be helpful.

I had a look at the failure here, and my gut feeling is that there could be some bugs in the existing code. It is failing on the "ClearControl" command first, trying to decrypt a set of parameters. The code assumes at least one parameter exists, but in this case there are zero parameters. If I hack around that, then there's a second very similar problem dealing with encrypted parameters again which I didn't understand well enough to debug further.

I did find our other unrelated bug when I added tests, which I've fixed in a separate commit in the new set of patches.

berrange commented:
> I did find our other unrelated bug when I added tests, which I've fixed in a separate commit in the new set of patches.

Oh nevermind, I see you also just fixed the same problem with the constants meanwhile.

joholl reviewed (approved):
LGTM. Thanks for your effort!
